e2e/mapimachines_test.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	configv1 "github.com/openshift/api/config/v1"
+	machinev1 "github.com/openshift/api/machine/v1beta1"
+	"github.com/openshift/cluster-capi-operator/e2e/framework"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/komega"
+)
+
+const (
+	machineAPINamespace = "openshift-machine-api"
+	clusterAPINamespace = "openshift-cluster-api"
+	RoleLabel           = "machine.openshift.io/cluster-api-machine-role"
+	DefaultTimeout      = 5 * time.Minute
+	DefaultInterval     = 10 * time.Second
+	// MAPI condition types
+	MapiConditionPaused   machinev1.ConditionType = "Paused"
+	ConditionSynchronized machinev1.ConditionType = "Synchronized"
+	// CAPI condition types
+	CapiConditionPaused = "Paused"
+)
+
+var _ = Describe("[sig-cluster-lifecycle][OCPFeatureGate:MachineAPIMigration] Machines Migration Tests", Ordered, func() {
+	BeforeAll(func() {
+		if platform != configv1.AWSPlatformType {
+			Skip(fmt.Sprintf("Skipping tests on %s, this only support on aws", platform))
+		}
+
+		if !framework.IsMachineAPIMigrationEnabled(ctx, cl) {
+			Skip("Skipping, this feature is only supported on MachineAPIMigration enabled clusters")
+		}
+	})
+
+	var _ = Describe("Mapi machine creation E2E", Ordered, func() {
+		var newMachine *machinev1.Machine
+		// This cleanup runs after each test spec.
+		AfterEach(func() {
+			By("Cleaning up created machine")
+			// If a new machine was created in the test, delete it.
+			if newMachine != nil {
+				err := cl.Delete(context.Background(), newMachine)
+				Expect(client.IgnoreNotFound(err)).To(Succeed(), "Failed to delete the new machine")
+				// Wait for the machine to be fully deleted.
+				Eventually(func() error {
+					key := client.ObjectKey{Name: newMachine.Name, Namespace: newMachine.Namespace}
+					err := cl.Get(context.Background(), key, &machinev1.Machine{})
```

Comment:
This will return nil when the object still exists, so this eventually isn't actually doing what you want it to, right?

I think you can instead return the error directly from `cl.Get` and then use `Should(WithTransform(apierrors.IsNotFound, BeTrue()), "Eventually the machine should not be found")`